v0/destinations/ga4/transform.js

```diff
@@ -0,0 +1,414 @@
+const get = require("get-value");
+const logger = require("../../../logger");
+const { EventType } = require("../../../constants");
+const {
+  CustomError,
+  defaultPostRequestConfig,
+  constructPayload,
+  defaultRequestConfig,
+  extractCustomFields,
+  isEmptyObject,
+  flattenJson,
+  getDestinationExternalID
+} = require("../../util");
+const {
+  ENDPOINT,
+  trackCommonConfig,
+  eventNameMapping,
+  mappingConfig,
+  ConfigCategory
+} = require("./config");
+const {
+  msUnixTimestamp,
+  isReservedEventName,
+  GA4_RESERVED_PARAMETER_EXCLUSION,
+  removeReservedParameterPrefixNames,
+  GA4_RESERVED_USER_PROPERTY_EXCLUSION,
+  removeReservedUserPropertyPrefixNames,
+  isReservedWebCustomEventName,
+  isReservedWebCustomPrefixName,
+  getDestinationItemProperties
+} = require("./utils");
+
+const trackResponseBuilder = async (message, { Config }) => {
+  let event = get(message, "event");
+  if (!event) {
+    throw new CustomError("Event name is required", 400);
+  }
+
+  // trim and replace spaces with "_"
+  event = event.trim().replace(/\s+/g, "_");
+
+  // reserved event names are not allowed
+  if (isReservedEventName(event)) {
+    throw new CustomError(
+      "[Google Analytics 4] track:: Reserved event names are not allowd",
+      400
+    );
+  }
+
+  // get common top level rawPayload
+  let rawPayload = constructPayload(message, trackCommonConfig);
+  if (rawPayload.timestamp_micros) {
+    rawPayload.timestamp_micros = msUnixTimestamp(rawPayload.timestamp_micros);
+  }
+
+  if (Config.measurementId) {
+    // gtag.js
+    if (!rawPayload.client_id) {
+      throw new CustomError(
+        "context.client_id or messageId must be provided",
+        400
+      );
+    }
+  } else {
+    // firebase
+    rawPayload.app_instance_id = getDestinationExternalID(
+      message,
+      "ga4AppInstanceId"
+    );
+    if (!rawPayload.app_instance_id) {
+      throw new CustomError(
+        "ga4AppInstanceId must be provided under externalId",
+        400
+      );
+    }
+    delete rawPayload.client_id;
+  }
+
+  const payload = {};
+  if (eventNameMapping[event.toLowerCase()]) {
+    // GA4 standard events
+    // get event specific parameters
+    switch (event.toLowerCase()) {
+      /* Browsing Section */
+      case "products_searched":
```

Comment:
Based on this ref:https://www.rudderstack.com/docs/rudderstack-api/api-specification/rudderstack-ecommerce-events-specification/ordering/

We shouldn't consider only snake case event naming here

Reply:
We are bringing strings to a common ground by trimming and replacing spaces with `_`. [L:40](https://github.com/rudderlabs/rudder-transformer/pull/1130/files/2b1d183dbf6234b61d1a047e1285e3259e74987c#diff-38fd1bc16aadc482899b425f3baf343c7ccfc49e7a9f0b794bae8f79a4ffaf04R40)